packages/sitecore-jss/src/personalization/layout-personalization-service.ts

```diff
@@ -0,0 +1,220 @@
+    let personalizedFragments: { [key: string]: ComponentRendering | null | undefined } = {};
+
+    try {
+      const personalizationDecisionsResult = await this.personalizationDecisionsService.getPersonalizationDecisions(
+        {
+          routePath: context.routePath,
+          language: context.language,
+          renderingIds: personalizedRenderingIds,
+        }
+      );
+      personalizedFragments = await this.resolveFragments(personalizationDecisionsResult, context);
+    } catch (error) {
+      // catch all errors on getting a personalization decision
+      console.error(error);
+      // default will be used for unresolved fragments
+      personalizedRenderingIds.forEach((id) => (personalizedFragments[id] = undefined));
+    }
+
+    const result: { [key: string]: ComponentRendering | null } = {};
+    personalizableRenderings.forEach((pr) => {
+      result[pr.uid] = this.layoutPersonalizationUtils.buildPersonalizedFragment(
+        pr.uid,
+        personalizedFragments,
+        pr.personalization.defaultComponent
+      );
+    });
+
+    return result;
+  }
+
+  /**
+   * Resolves the fragments.
+   * @param {PersonalizationDecisionData} personalizationDecisionsResult The personalization decisions.
+   * @param {PersonalizationContext} context The context.
+   * @returns {Object.<string, ComponentRendering | null | undefined>} The fragments.
+   */
+  private async resolveFragments(
+    personalizationDecisionsResult: PersonalizationDecisionData,
+    context: PersonalizationContext
+  ): Promise<{ [key: string]: ComponentRendering | null | undefined }> {
+    const personalizedFragments: { [key: string]: ComponentRendering | null | undefined } = {};
+    const renderingsDecisions = personalizationDecisionsResult.renderings;
+    const personalizedFragmentsRequests: Promise<void>[] = [];
+
+    for (const [renderingId, decision] of Object.entries(renderingsDecisions)) {
+      const variantKey = decision?.variantKey;
+      if (variantKey) {
+        // load fragments in parallel
+        personalizedFragmentsRequests.push(
+          this.layoutFragmentService
+            .fetchLayoutFragmentData(context.routePath, context.language, renderingId, variantKey)
+            .then((fr) => {
+              personalizedFragments[renderingId] = fr.fragment;
+            })
+            .catch((error) => {
+              console.error(error);
+
+              // default will be used in case failed to resolve the fragment
+              personalizedFragments[renderingId] = undefined;
+            })
+        );
+      } else if (variantKey === null) {
+        // hidden by personalization
+        personalizedFragments[renderingId] = null;
+      } else {
+        // was not able to resolve decision for the rendering, default will be used
+        personalizedFragments[renderingId] = undefined;
+      }
+    }
+
+    await Promise.all(personalizedFragmentsRequests);
```

Comment:
Do we need to await here? Could we just return the promise and let the consumer decide whether to wait for the result

Reply:
We need to await personalizedFragmentsRequests as personalizedFragments is build on these tasks.
await is not blocking as long as somebody does not await for the result. The purpose of async/await is to simplify the syntax necessary to consume promise-based APIs.